### PR TITLE
Add hyper::Uri support

### PR DESCRIPTION
**Pull request by ferranpujolcamins**
Based on PR https://github.com/nox/hyper_serde/pull/20

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -17,12 +17,15 @@ doctest = false
...
 serde = "1.0"
 serde_bytes = "0.10"
 time = "0.1"
 
 [dev-dependencies]
 serde_test = "1.0"
 time = "0.1"
+
+[replace]
+"hyper:0.11.2" = { git = "https://github.com/hyperium/hyper.git" }
```
Comment (nox) [inline Cargo.toml]: Why is this here?

File: src/lib.rs [hunk 2]
```diff
@@ -412,7 +415,7 @@ impl<'a> Serialize for Ser<'a, Headers> {
         for header in self.v.iter() {
             let name = header.name();
             let value = self.v.get_raw(name).unwrap();
-            serializer.serialize_entry(name, &Value(value, self.pretty))?;
+            serializer.serialize_entry(name, &Value(&value.iter().map(|v| v.to_vec()).collect::<Vec<Vec<u8>>>(), self.pretty))?;
```
Comment (nox) [inline src/lib.rs]: Why does this need to allocate a new `Vec<Vec<u8>>`?

File: src/lib.rs [hunk 3]
```diff
@@ -534,3 +537,37 @@ impl<'a> Serialize for Ser<'a, Tm> {
...
+            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+                write!(formatter, "an HTTP Uri value")
+            }
+
+            fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>
+                where E: de::Error,
+            {
+                Uri::from_str(v)
+                    .map(De::new)
+                    .map_err(|e| E::custom(format!("{:?}", e)))
```
Comment (nox) [inline src/lib.rs]: That shouldn't use `Debug`.

File: src/lib.rs [hunk 4]
```diff
@@ -534,3 +537,37 @@ impl<'a> Serialize for Ser<'a, Tm> {
...
+
+        deserializer.deserialize_string(UriVisitor)
+    }
+}
+
+impl<'a> Serialize for Ser<'a, Uri> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+        where S: Serializer,
+    {
+        serializer.serialize_str(&self.v.to_string())
```
Comment (nox) [inline src/lib.rs]: This allocates for nothing.

File: Cargo.toml [hunk 5]
```diff
@@ -17,12 +17,15 @@ doctest = false
 
 [dependencies]
 cookie = {version = "0.6", default-features = false}
-hyper = "0.10"
-mime = "0.2"
+hyper = { version = "0.11", features = ["raw_status"] }
```
Comment (nox) [inline Cargo.toml]: Why is `raw_status` needed?

**Comment by DarrenTsung** [comment]
I opened another PR from my fork integrating the feedback, see #23.

**Comment by ferranpujolcamins** [comment]
This PR was erased from my memory. Thanks @DarrenTsung for picking it up!
